src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                        if (subsumedErrorCode != 0 && subsumedErrorCode != ErrorCode.ERR_NoImplicitConvCast)
+                        {
+                            if (!label.HasErrors)
+                            {
+                                _diagnostics.Add(subsumedErrorCode, label.Pattern.Syntax.Location);
+                            }
+                        }
+                        else
+                        {
+                            AddToDecisionTree(result, label.Pattern, label.Guard, label);
+                        }
+                    }
+                }
+            }
+
+            if (defaultLabel != null)
+            {
+                Add(result, (e, t) => new DecisionTree.Guarded(_switchStatement.Expression, _switchStatement.Expression.Type, default(ImmutableArray<KeyValuePair<BoundExpression, LocalSymbol>>), null, defaultLabel));
+            }
+
+            return result;
+        }
+
+        private void AddToDecisionTree(DecisionTree decisionTree, BoundPattern pattern, BoundExpression guard, BoundPatternSwitchLabel label)
+        {
+            if (guard?.ConstantValue == ConstantValue.False) return;
+            switch (pattern.Kind)
+            {
+                case BoundKind.ConstantPattern:
+                    {
+                        var constantPattern = (BoundConstantPattern)pattern;
+                        AddByValue(decisionTree, constantPattern.Value, (e, t) => new DecisionTree.Guarded(e, t, default(ImmutableArray<KeyValuePair<BoundExpression, LocalSymbol>>), guard, label), label.HasErrors);
+                        break;
+                    }
+                case BoundKind.DeclarationPattern:
+                    {
+                        var declarationPattern = (BoundDeclarationPattern)pattern;
+                        DecisionMaker maker = (e, t) => new DecisionTree.Guarded(e, t, ImmutableArray.Create(new KeyValuePair<BoundExpression, LocalSymbol>(e, declarationPattern.LocalSymbol)), guard, label);
+                        if (declarationPattern.IsVar)
+                        {
+                            Add(decisionTree, maker);
+                        }
+                        else
+                        {
+                            AddByType(decisionTree, declarationPattern.DeclaredType.Type, maker);
+                        }
+                        break;
+                    }
+                case BoundKind.WildcardPattern:
+                    // We do not yet support a wildcard pattern syntax. It is used exclusively
+                    // to model the "default:" case, which is handled specially in the caller.
+                    throw ExceptionUtilities.UnexpectedValue(pattern.Kind);
+            }
+        }
+
+        private ErrorCode CheckSubsumed(BoundPattern pattern, DecisionTree decisionTree, bool inputCouldBeNull)
+        {
+            if (decisionTree.MatchIsComplete) return ErrorCode.ERR_PatternIsSubsumed;
+            switch (pattern.Kind)
```

Comment:
I started out that way, found it to be too messy, and moved to this. That would be an anti-pattern we've carefully avoided throughout the compilers. A decision tree is akin to a bound node, and we do not add virtual methods to bound nodes every place we might switch on its kind.

---
In reply to: [64855879](https://github.com/dotnet/roslyn/pull/11485#discussion_r64855879) [](ancestors = 64855879)